### ISSUE TITLE
opal-builder.js (0.11) is producing invalid JavaScript

### ISSUE THREAD

**Issue by Mogztter** [opal/opal#1636]
When trying to compile Asciidoctor to JavaScript using opal-builder.js, I get the following error at runtime:

```
/home/guillaume/workspace/opensource/asciidoctor/asciidoctor.js/build/asciidoctor.js:21733
        return ($a = transform, $b = self.converter_map, ((($c = $b['$[]']($a)) !== false && $c !== nil && $c != null) ? $c : $b['$[]=']($a, self.$find_converter(transform))))((($a = ) !== false && $a !== nil && $a != null) ? $a : self.converter_map['$[]'](transform, self.$find_converter(transform)))
                                                                                                                                                                                       ^

SyntaxError: Unexpected token )
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:374:25)
    at Object.Module._extensions..js (module.js:417:10)
    at Module.load (module.js:344:32)
    at Function.Module._load (module.js:301:12)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/home/guillaume/workspace/opensource/asciidoctor/asciidoctor.js/build/bower.spec.all.js:311:19)
    at Module._compile (module.js:410:26)
    at Object.Module._extensions..js (module.js:417:10)
```

Here's the line in Ruby: https://github.com/asciidoctor/asciidoctor/blob/v1.5.5/lib/asciidoctor/converter/composite.rb#L41

@mojavelinux made a change (https://github.com/asciidoctor/asciidoctor/commit/19cce5c57db3056ee51ab55facbfb5c65fa085f5) to workaround a parse bug but maybe this change is not necessary anymore and we should remove these parenthesis ? 

Never the less, I think both syntax should be valid no ?

Ref: https://github.com/asciidoctor/asciidoctor.js/issues/279

**Comment by Mogztter** [comment]
@elia Is it possible to let the user decide if they want to use the "old" self-written lexer/parser or the [new parser](https://github.com/whitequark/parser/) ?
I'm asking because I don't think I will be able to make it as fast as the old one. Plus I was pretty satisfy with the self-written one :wink: 
If this is not possible, my only alternative will be to maintain a fork but I don't want to do that :neutral_face:

**Comment by iliabylich** [comment]
@Mogztter I think it's impossible atm. New parser produces a bit different AST which requires some serious support in `opal/nodes` to work with different versions of nodes. Moreover, we depend on some parser features like AST rewriters.

Does asciidoctor (Ruby version) use `Kernel#eval(code)` (or `Kernel#module_eval(code)`)? Does asciidoctor.js uses `Opal.compile` and other lexing/parsing/compilation stuff in runtime? Because if you use parser only to compile initial ruby code to js (asciidoctor.rb to asciidoctor.js) - then most probably it's not affected by switching to https://github.com/whitequark/parser. I remember there were a few intended performance degradations (like stack-based constant lookup and post-args support), but all of them were minor and they were (and still are) necessary for compatibility with MRI.

**Comment by Mogztter** [comment]
> Does asciidoctor (Ruby version) use Kernel#eval(code) (or Kernel#module_eval(code))?

No I don't think so.

> Does asciidoctor.js uses Opal.compile and other lexing/parsing/compilation stuff in runtime? Because if you use parser only to compile initial ruby code to js (asciidoctor.rb to asciidoctor.js) - then most probably it's not affected by switching to https://github.com/whitequark/parser. I remember there were a few intended performance degradations (like stack-based constant lookup and post-args support), but all of them were minor and they were (and still are) necessary for compatibility with MRI.

We are using `whitequark/parser` (compiled in JavaScript) to compile Asciidoctor Ruby to JavaScript. In other word Asciidoctor.js build relies on `whitequark/parser`.
The thing is that `whitequark/parser` is slower than the self-written lexer/parser (at least when compiled in JavaScript). 
With Opal 0.10.3, the compilation of Asciidoctor Ruby takes 8 seconds. With Opal master, the compilation takes 180 seconds (that's more than 20 times slower!).

I didn't have the chance to compare the runtime performance of Asciidoctor.js because of this issue but I'm not too worried.

**Comment by elia** [comment]
@Mogztter as @iliabylich said there's no way we keep both around, would be easier to optimize opal and mitigate the difference but given how much more complex a complete Ruby parser is I don't think we'll go back to 8s easily. 

My advice would be to precompile Asciidoctor with Ruby and ship it as a js, but admittedly I don't remember what was the target audience for the live compilation.

---

All that said the one in the original post seems a genuine bug 🐛

**Comment by Mogztter** [comment]
> as @iliabylich said there's no way we keep both around, would be easier to optimize opal and mitigate the difference but given how much more complex a complete Ruby parser is I don't think we'll go back to 8s easily.

I do think this is the right move here and I will try to keep improving the performance 😉 

> My advice would be to precompile Asciidoctor with Ruby and ship it as a js, but admittedly I don't remember what was the target audience for the live compilation.

Ideally, I want to be able to build Asciidoctor.js in JavaScript because I don't want to "scare" a potential contributor with a build that requires Ruby in a JavaScript project 😆

**Comment by Mogztter** [comment]
I could not reproduce this issue when using Opal builder in Ruby

**Comment by mojavelinux** [comment]
> I want to be able to build Asciidoctor.js in JavaScript because I don't want to "scare" a potential contributor with a build that requires Ruby in a JavaScript project

I agree it's that, but it's also more. It's about integrating into the community with a native toolchain. A JavaScript project should be built with JavaScript. Otherwise, it's always an outsider.

**Comment by mojavelinux** [comment]
> @Mogztter wrote: With Opal 0.10.3, the compilation of Asciidoctor Ruby takes 8 seconds. With Opal master, the compilation takes 180 seconds (that's more than 20 times slower!).

Ouch. That's quite a difference. Since this is compilation, some performance degradation is acceptable if it means more accuracy, but we can hopefully be in the same order of magnitude.

**Comment by mojavelinux** [comment]
> Does asciidoctor (Ruby version) use Kernel#eval(code) (or Kernel#module_eval(code))?

No, Asciidoctor tries to be as vanilla as possible. I understand the risks of too much dynamic behavior, so I mostly avoided it.

The only dynamic calls are in the extensions, which use instance_exec in some cases.

**Comment by Mogztter** [comment]
I've created a simple project to reproduce this issue: https://github.com/Mogztter/opal-issue-1636


#### Minimal Ruby code

```ruby
@converter_map = {}

def find_converter transform
end

def converter_for transform
  @converter_map[transform] ||= (find_converter transform)
end
```

#### Compile in Ruby
I'm using Opal and Ruby to compile this code:

```
bundle install
bundle exec ruby compile.rb
```

`compile.rb` is using Opal builder:

```ruby
require 'opal'

builder = Opal::Builder.new
builder.append_paths('.')
builder.build('code.rb')
File.write 'build/result-ruby.js', builder.to_s
```

#### Compile in JavaScript

I'm using Node.js and npm:

```
npm install
node compile.js
```

`compile.js` is using [Opal Node Compiler](https://github.com/mogztter/opal-node-compiler):

```
var fs = require('fs');
var Builder = require('opal-compiler').Builder;

var builder = Builder.$new();
var result = builder.$build('code.rb').$to_s();
fs.writeFileSync('build/result-js.js', result, 'utf8');
```

#### Compare

Both scripts compile and write the result in the `build` directory. You can see that `result-ruby.js` is invalid but `result-js.js` is not.

**Comment by Mogztter** [comment]
I'm making progress.
I'm using the following code to make the comparison easier:

```ruby
def converter_for transform
  @map[transform] ||= (find_converter transform)
end
```

The AST produced by `parser` in Ruby and in JavaScript are identical but when we process this AST with Opal, there's one difference in [compiler#process()](https://github.com/opal/opal/blob/299c8cdc00124ae9a561ee288486b6ce88c3c9f3/lib/opal/compiler.rb#L339).

In Ruby, the code `(find_converter transform)` is translated into `send, self, arglist, lvar, arg` but in JavaScript `send, self, arglist, lvar, or, ...` (notice the `or` instead of `arg`)

**Comment by Mogztter** [comment]
Gotcha !!!

**Comment by Mogztter** [comment]
How to reproduce this issue:

```ruby
def compile type
  get_node = case type
  when :lvasgn then 'lvasgn'
  when :send
    p 'send'
    return
  else
    raise "Unsupported node in LogicalOpAssignNode #{type}"
  end

  p 'rest'
  p get_node
end

compile :send
```
MRI:
```
bundle exec ruby test.rb   
"send"
```
Opal:
```
$ bundle exec opal -- test.rb 
"send"
"rest"
nil
```

**Comment by Mogztter** [comment]
The issue seems to be related to the assignment of the switch/case.
The following is working:
```ruby
def compile type
  case type
  when :lvasgn then 'lvasgn'
  when :send
    p 'send'
    return
  else
    raise "Unsupported node in LogicalOpAssignNode #{type}"
  end

  p 'rest'
end
```
The following is not working:
```ruby
def compile type
  get_node = case type
  when :lvasgn then 'lvasgn'
  when :send
    p 'send'
    return
  else
    raise "Unsupported node in LogicalOpAssignNode #{type}"
  end

  p 'rest'
end
```

**Comment by Mogztter** [comment]
@elia In this case we should use `throw $breaker;` right ?

**Comment by Mogztter** [comment]
For now I've just rewrote [this code in call_special.rb](https://github.com/opal/opal/blob/299c8cdc00124ae9a561ee288486b6ce88c3c9f3/lib/opal/nodes/call_special.rb#L69-L83) to not use a `return` in a switch / case

**Comment by Mogztter** [comment]
The root cause is not resolved but with this commit 26ac424, `opal-builder.js` is now valid :tada:
